src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,433 @@
+                @Override
+                public boolean apply(JavaField input) {
+                    return ZERO_SIZE_ARRAY_FIELDS.contains(input.getFullName());
+                }
+            };
+
+        final ArchCondition<JavaField> beSuppressed = new ArchCondition<>("be suppressed") {
+            @Override
+            public void check(JavaField item, ConditionEvents events) {
+                if (!SUPPRESSED_FIELDS_IN_UTIL_CLASSES.contains(item.getFullName())) {
+                    final String message = String
+                        .format(Locale.ROOT, "should %s or resolved.", getDescription());
+                    events.add(SimpleConditionEvent.violated(item, message));
+                }
+            }
+        };
+
+        final ArchRule fieldsShouldBeImmutable = fields()
+            .that(are(not(zeroSizeArrays)))
+            .and()
+            .areDeclaredInClassesThat()
+            .haveSimpleNameEndingWith("Util")
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould()
+            .beStatic()
+            .orShould(beSuppressed);
+
+        fieldsShouldBeImmutable.check(utilClasses);
+    }
+
+    /**
+     * Test to ensure modules annotated with {@link StatelessCheck} contain immutable fields
+     * and modules with mutable fields are annotated with {@link FileStatefulCheck} or
+     * {@link GlobalStatefulCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testModuleImmutability() {
+        final JavaClasses checkstyleChecks = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.checks");
+
+        final ModulePropertyPredicate moduleProperties = new ModulePropertyPredicate();
+
+        final DescribedPredicate<JavaClass> immutableFields = new ImmutableFieldsPredicate();
+
+        final ArchCondition<JavaField> bePrivateExceptMessageKeys =
+            new ArchCondition<>("be private") {
+
+                @Override
+                public void check(JavaField item, ConditionEvents events) {
+                    if (!item.getName().startsWith("MSG")
+                        && !item.getModifiers().contains(JavaModifier.PRIVATE)) {
+                        final String message = String
+                            .format(Locale.ROOT, "Field <%s> should %s in %s",
+                                    item.getFullName(), getDescription(),
+                                    item.getSourceCodeLocation().toString());
+                        events.add(SimpleConditionEvent.violated(item, message));
+                    }
+                }
+            };
+
+        final ArchCondition<JavaClass> beAnnotatedOrSuppressed = new ArchCondition<>(
```

Comment:
I do not see logic here for annotation checking, what am I missing?

Reply:
Nothing, the variable name was incorrect, it only checks suppressions.